api-authentication/server/index.js

```diff
@@ -0,0 +1,115 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+  const params = [username];
+  const sql = `
+  select "userId", "hashedPassword"
+  from "users"
+  where "username" = $1
+  `;
+  db.query(sql, params)
+    .then(result => {
+      if (result.rows.length < 1) {
+        throw new ClientError(404, 'invalid login error');
+      }
+      // else {
+      // res.status(201).json(result.rows);
+      // }
+      argon2
+        .verify(result.rows[0].hashedPassword, password)
+        .then(isMatching => {
+          // console.log('Does your password match?', isMatching);
+          if (isMatching) {
+            const payload = {
+              userId: result.rows[0].userId,
+              username:
+                username
+            };
+
+            const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+
+            res.status(201).json({ payload: payload, token: token });
+
+          } else {
+            throw new ClientError(401, 'Password did not match');
+          }
+        })
+        .catch(err => {
+          console.error(err);
+        });
+    })
```

Comment:
Be sure to also `.catch()` at the end of your `db.query()` promise chain in case querying the database rejects. Otherwise the client will hand and time out while your server leaks a socket connection.

Reply:
![Screen Shot 2022-04-06 at 10 13 37 AM](https://user-images.githubusercontent.com/40590740/162030976-895941f4-dd08-4024-bf19-8d5791f43552.png)


Hi I think I did, it was just under my pseudo code.